core/engine/src/builtins/date/utils.rs

```diff
@@ -1,202 +1,730 @@
+///
+/// [spec]: https://tc39.es/ecma262/#sec-timezoneestring
+pub(super) fn time_zone_string(t: f64, hooks: &dyn HostHooks) -> JsString {
+    // 1. Let systemTimeZoneIdentifier be SystemTimeZoneIdentifier().
+    // 2. If IsTimeZoneOffsetString(systemTimeZoneIdentifier) is true, then
+    //     a. Let offsetNs be ParseTimeZoneOffsetString(systemTimeZoneIdentifier).
+    // 3. Else,
+    //     a. Let offsetNs be GetNamedTimeZoneOffsetNanoseconds(systemTimeZoneIdentifier, ℤ(ℝ(tv) × 10**6)).
+    // 4. Let offset be 𝔽(truncate(offsetNs / 10**6)).
+    let offset = f64::from(local_timezone_offset_seconds(t, hooks)) * MS_PER_SECOND;
+    //let offset = hooks.local_timezone_offset_seconds((t / MS_PER_SECOND).floor() as i64);
+
+    // 5. If offset is +0𝔽 or offset > +0𝔽, then
+    let (offset_sign, abs_offset) = if offset >= 0.0 {
+        // a. Let offsetSign be "+".
+        // b. Let absOffset be offset.
+        (utf16!("+"), offset)
+    }
+    // 6. Else,
+    else {
+        // a. Let offsetSign be "-".
+        // b. Let absOffset be -offset.
+        (utf16!("-"), -offset)
     };
 
-    let new_day = make_day(year, month, date)?;
-    let new_time = make_time(hour, minute, second, millisecond)?;
-    let mut ts = make_date(new_day, new_time)?;
+    // 7. Let offsetMin be ToZeroPaddedDecimalString(ℝ(MinFromTime(absOffset)), 2).
+    let offset_min = pad_two(min_from_time(abs_offset));
+
+    // 8. Let offsetHour be ToZeroPaddedDecimalString(ℝ(HourFromTime(absOffset)), 2).
+    let offset_hour = pad_two(hour_from_time(abs_offset));
+
+    // 9. Let tzName be an implementation-defined string that is either the empty String or the
+    // string-concatenation of the code unit 0x0020 (SPACE), the code unit 0x0028 (LEFT PARENTHESIS),
+    // an implementation-defined timezone name, and the code unit 0x0029 (RIGHT PARENTHESIS).
+    // 10. Return the string-concatenation of offsetSign, offsetHour, offsetMin, and tzName.
+    js_string!(offset_sign, &offset_hour, &offset_min)
+}
+
+/// Abstract operation `ToDateString ( tv )`
+///
+/// More info:
+/// - [ECMAScript reference][spec]
+///
+/// [spec]: https://tc39.es/ecma262/#sec-todatestring
+pub(super) fn to_date_string_t(tv: f64, hooks: &dyn HostHooks) -> JsString {
+    // 1. If tv is NaN, return "Invalid Date".
+    if tv.is_nan() {
+        return JsString::from("Invalid Date");
+    }
 
-    if LOCAL {
-        ts = hooks
-            .local_from_naive_local(NaiveDateTime::from_timestamp_millis(ts)?)
-            .earliest()?
-            .naive_utc()
-            .timestamp_millis();
+    // 2. Let t be LocalTime(tv).
+    let t = local_time(tv, hooks);
+
+    // 3. Return the string-concatenation of
+    // DateString(t),
+    // the code unit 0x0020 (SPACE),
+    // TimeString(t),
+    // and TimeZoneString(tv).
+    js_string!(
+        &date_string(t),
+        utf16!(" "),
+        &time_string(t),
+        &time_zone_string(t, hooks)
+    )
+}
+
+fn modulo(n: f64, m: f64) -> f64 {
```

Comment:
question: is there any benefit/difference using this modulo function vs. `rem_euclid`